downstairs/src/extent_inner_raw_v2.rs

```diff
@@ -0,0 +1,1058 @@
+            false, // dirty
+            new_flush,
+            new_gen,
+        )?;
+        self.dirty = false;
+        Ok(())
+    }
+
+    fn get_recordsize(path: &Path) -> Result<u64, CrucibleError> {
+        let recordsize = {
+            let p = std::process::Command::new("zfs")
+                .arg("get")
+                .arg("-Hp") // scripting mode
+                .arg("-ovalue")
+                .arg("recordsize")
+                .arg(path)
+                .stdout(std::process::Stdio::piped())
+                .stderr(std::process::Stdio::piped())
+                .spawn();
+            match p {
+                Ok(mut p) => {
+                    p.wait().map_err(|e| {
+                        CrucibleError::IoError(format!(
+                            "call to `zfs` failed: {e}"
+                        ))
+                    })?;
+                    let mut err = vec![];
+                    p.stderr.unwrap().read_to_end(&mut err).map_err(|e| {
+                        CrucibleError::IoError(format!(
+                            "failed to read stderr from `zfs`: {e:?}"
+                        ))
+                    })?;
+                    let err = std::str::from_utf8(&err).map_err(|e| {
+                        CrucibleError::IoError(format!(
+                            "zfs returned invalid UTF-8 string: {err:?} ({e})"
+                        ))
+                    })?;
+                    if err.contains("not a ZFS filesystem") {
+                        DEFAULT_ZFS_RECORDSIZE
+                    } else {
+                        let mut out = vec![];
+                        p.stdout.unwrap().read_to_end(&mut out).map_err(
+                            |e| {
+                                CrucibleError::IoError(format!(
+                                    "failed to read stdout from `zfs`: {e:?}, \
+                                     stderr: {err}"
+                                ))
+                            },
+                        )?;
+                        let out = std::str::from_utf8(&out).map_err(|e| {
+                            CrucibleError::IoError(format!(
+                                "zfs returned invalid UTF-8 string: {out:?} \
+                                 ({e}), stderr: {err}"
+                            ))
+                        })?;
+                        out.trim().parse::<u64>().map_err(|e| {
+                            CrucibleError::IoError(format!(
+                                "zfs returned non-integer for recordsize: \
+                                  {out:?} ({e}), stderr: {err}"
+                            ))
+                        })?
+                    }
+                }
+                Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
+                    // If the `zfs` executable isn't present, then we're
+                    // presumably on a non-ZFS filesystem and will use a default
+                    // recordsize
```

Comment:
There's a remote possibility that PATH isn't set correctly but we are on a ZFS filesystem. I'm wondering if the downstairs should somehow expect that this binary is present if we're running on illumos (we already expect it for taking snapshots!)

Reply:
Good idea, changed in 5c446c791dac to return an error if `zfs` isn't present on an illumos system.